File: core/src/main/java/heroes/journey/entities/tagging/Group.java
package heroes.journey.entities.tagging;

import java.util.Objects;

import heroes.journey.modlib.attributes.IGroup;
import heroes.journey.modlib.registries.Registrable;
import heroes.journey.mods.Registries;

public class Group extends Registrable implements IGroup {

    public Group(String id) {
        super(id);
    }

    @Override
    public Group register() {
        return Registries.GroupManager.register(this);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        Group group = (Group)o;
        return getId().equals(group.getId());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getId());
    }
}

File: modlib/src/main/java/heroes/journey/modlib/attributes/Group.kt
package heroes.journey.modlib.attributes

import heroes.journey.modlib.registries.Registrable
import heroes.journey.modlib.registries.Registries

/**
 * A Group, used for stat grouping and categorization.
 * This is a simple data container with no complex functions.
 */
class Group(id: String) : Registrable(id) {

    override fun register(): Group {
        Registries.GroupManager.register(this)
        return this
    }
}

/**
 * Builder for defining a group in a natural DSL style.
 */
class GroupBuilder {
    var id: String = ""
}

/**
 * DSL entrypoint for defining a group using a builder lambda.
 *
 * Example usage:
 * ```kotlin
 * group {
 *     id = Ids.GROUP_BODY
 * }
 * ```
 */
fun group(init: GroupBuilder.() -> Unit): Group {
    val builder = GroupBuilder()
    builder.init()
    return Group(builder.id)
}
